test/python/circuit/library/test_adders.py

```diff
@@ -0,0 +1,107 @@
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Test adder circuits."""
+
+import unittest
+import numpy as np
+from ddt import ddt, data, unpack
+
+from qiskit.test.base import QiskitTestCase
+from qiskit.circuit import QuantumCircuit
+from qiskit.quantum_info import Statevector
+from qiskit.circuit.library import RippleCarryAdder, QFTAdder, ClassicalAdder
+
+
+@ddt
+class TestAdder(QiskitTestCase):
+    """Test the adder circuits."""
+
+    def assertAdditionIsCorrect(self,
+                                num_state_qubits: int,
+                                adder: QuantumCircuit,
+                                inplace: bool,
+                                modular: bool):
+        """Assert that adder correctly implements the summation.
+
+        Args:
+            num_state_qubits: The number of bits in the numbers that are added.
+            adder: The circuit performing the addition of two numbers with ``num_state_qubits``
+                bits.
+            inplace: If True, compare against an inplace addition where the result is written into
+                the second register plus carry qubit. If False, assume that the result is written
+                into a third register of appropriate size.
+            modular: If True, omit the carry qubit to obtain an addition modulo
+                ``2^num_state_qubits``.
+        """
+        circuit = QuantumCircuit(*adder.qregs)
+
+        # create equal superposition
+        circuit.h(range(2 * num_state_qubits))
+
+        # apply adder circuit
+        circuit.compose(adder, inplace=True)
+
+        # obtain the statevector and the probabilities, we don't trace out the ancilla qubits
+        # as we verify that all ancilla qubits have been uncomputed to state 0 again
+        statevector = Statevector(circuit)
+        probabilities = np.abs(statevector) ** 2
+        pad = '0' * circuit.num_ancillas  # state of the ancillas
+
+        # compute the expected results
+        expectations = np.zeros_like(probabilities)
+        num_bits_sum = num_state_qubits + 1
+        # iterate over all possible inputs
+        for x in range(2 ** num_state_qubits):
+            for y in range(2 ** num_state_qubits):
+                # compute the sum
+                addition = (x + y) % (2 ** num_state_qubits) if modular else x + y
+                # compute correct index in statevector
+                bin_x = bin(x)[2:].zfill(num_state_qubits)
+                bin_y = bin(y)[2:].zfill(num_state_qubits)
+                bin_res = bin(addition)[2:].zfill(num_bits_sum)
+                bin_index = pad + bin_res + bin_x if inplace else pad + bin_res + bin_y + bin_x
+                index = int(bin_index, 2)
+                expectations[index] += 1 / 2 ** (2 * num_state_qubits)
+        np.testing.assert_array_almost_equal(expectations, probabilities)
```

Comment:
This test seems clever but I'm not sure I get all the details. So you put the circuit into a 2N-qubit equal superposition and run the adder on it, and expect the output would be the sum of all possible N-bit numbers?

Reply:
So the input registers `a` and `b` are in equal superposition, so the outcome is a equal superposition of all the possible additions with `N` bit numbers. This way we can check all possible additions at once, in a single statevector. Of course you could argue that this test succeeds for another operation than addition, e.g. a special operation that works only on equal superpositions 🤔

Reply:
ok maybe add a comment.